py-polars/polars/interchange/from_dataframe.py

```diff
@@ -37,41 +44,277 @@ def from_dataframe(df: SupportsInterchange, *, allow_copy: bool = True) -> DataF
+    if polars_dtype == Boolean and dtype[1] == 8:
+        if length == 0:
+            return pl.Series(dtype=Boolean)
+        elif not allow_copy:
+            msg = "byte-packed boolean buffer must be converted to bit-packed boolean"
+            raise CopyNotAllowedError(msg)
+        return pl.Series._from_buffer(UInt8, buffer_info, owner=buffer).cast(Boolean)
+
+    return pl.Series._from_buffer(polars_dtype, buffer_info, owner=buffer)
+
+
+def _construct_offsets_buffer(
+    buffer: Buffer,
+    dtype: Dtype,
+    offset: int,
+    *,
+    allow_copy: bool,
+) -> Series:
+    polars_dtype = dtype_to_polars_dtype(dtype)
+    length = get_buffer_length_in_elements(buffer.bufsize, dtype) - offset
+
+    buffer_info = (buffer.ptr, offset, length)
+    s = pl.Series._from_buffer(polars_dtype, buffer_info, owner=buffer)
+
+    # Polars only supports Int64 offsets
+    if polars_dtype != Int64:
+        if not allow_copy:
+            msg = f"offsets buffer must be cast from {polars_dtype} to Int64"
+            raise CopyNotAllowedError(msg)
+        s = s.cast(Int64)
+
+    return s
+
+
+def _construct_validity_buffer(
+    validity_buffer_info: tuple[Buffer, Dtype] | None,
+    column: Column,
+    column_dtype: PolarsDataType,
+    data: Series,
+    offset: int = 0,
+    *,
+    allow_copy: bool,
+) -> Series | None:
+    null_type, null_value = column.describe_null
+    if null_type == ColumnNullType.NON_NULLABLE or column.null_count == 0:
+        return None
+
+    elif null_type == ColumnNullType.USE_BITMASK:
+        if validity_buffer_info is None:
+            return None
+        buffer = validity_buffer_info[0]
+        return _construct_validity_buffer_from_bitmask(
+            buffer, null_value, column.size(), offset, allow_copy=allow_copy
         )
 
-    if isinstance(df, pa.Table):
-        return df
-    elif isinstance(df, pa.RecordBatch):
-        return pa.Table.from_batches([df])
+    elif null_type == ColumnNullType.USE_BYTEMASK:
+        if validity_buffer_info is None:
+            return None
+        buffer = validity_buffer_info[0]
+        return _construct_validity_buffer_from_bytemask(
+            buffer, null_value, allow_copy=allow_copy
+        )
+
+    elif null_type == ColumnNullType.USE_NAN:
+        if not allow_copy:
+            msg = "bitmask must be constructed"
+            raise CopyNotAllowedError(msg)
+        return data.is_not_nan()
+
+    elif null_type == ColumnNullType.USE_SENTINEL:
+        if not allow_copy:
+            msg = "bitmask must be constructed"
+            raise CopyNotAllowedError(msg)
+
+        try:
+            sentinel = pl.Series([null_value])
+            if column_dtype.is_temporal():
+                sentinel = sentinel.cast(column_dtype)
+            return data != sentinel  # noqa: TRY300
+        except ComputeError as e:
+            msg = f"invalid sentinel value for column of type {column_dtype}: {null_value!r}"
+            raise TypeError(msg) from e
```

Comment:
I think the ruff code here is a good one?
```python
        sentinel = pl.Series([null_value])
        if column_dtype.is_temporal():
            try:
                sentinel = sentinel.cast(column_dtype)
            except ComputeError as e:
                msg = f"invalid sentinel value for column of type {column_dtype}: {null_value!r}"
                raise TypeError(msg) from e
        return data != sentinel
```
so that the `try` block is as small and atomic as possible

Reply:
Both the `cast` and the equality operation can cause a ComputeError if the sentinel value is invalid. So I have them both in the try block. Placing the sentinel in a Series shouldn't fail, so I'll extract that one.